File: scripts/live-qa-transport.mjs
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const helper = fileURLToPath(new URL('./live_qa_guard.py', import.meta.url));
export const BASE_URL = 'https://api.conifer.build';
export const BUILD = 'e3fd202';
export const PINS = { chat: 'gpt-4.1-nano', spare: 'gpt-4.1-mini', native: 'claude-haiku-4-5', embed: 'text-embedding-3-small' };

export function journal(action, payload = {}) {
  const result = spawnSync(process.env.CONIFER_QA_PYTHON || 'python3', [helper], {
    input: JSON.stringify({ action, payload: { phase: 'typescript', ...payload } }), encoding: 'utf8',
  });
  if (result.error) throw result.error;
  const reply = JSON.parse(result.stdout || '{}');
  if (result.status !== 0) throw new Error(reply.guard_error || result.stderr || 'QA guard failed');
  return reply;
}

export function assertModel(model, cap, capability, provider) {
  if (!model || model.endpointKind !== 'conifer' || model.unavailable === true ||
      (capability && !model.caps?.includes(capability)) || (provider && model.provider !== provider)) {
    throw new Error('fixed model is absent, unavailable or incompatible with this wire');
  }
  for (const field of ['in_usd_per_mtok', 'out_usd_per_mtok']) {
    const price = model.pricing?.[field];
    if (typeof price !== 'string' || !/^\d+(?:\.\d+)?$/.test(price)) throw new Error('unpriced model');
  }
  if (cap !== undefined && (model.outputTokenLimitSupported === false ||
      (model.minOutputTokens ?? 0) > cap || (model.maxOutputTokens ?? cap) < cap)) {
    throw new Error(`${model.id} cannot honor ${cap} output tokens`);
  }
  return model;
}

export function emptyOutcome(error) {
  if (error?.status !== 422 || error?.type !== 'output_budget_exhausted' ||
      error?.code !== 'output_budget_exhausted' || error?.param !== 'max_tokens' || error?.retryable !== false ||
      !error.requestId || !error.body?.usage) throw error;
  return 'typed output_budget_exhausted with preserved usage; transport verified settled receipt';
}

export function createGuardedFetch({ dispatch = globalThis.fetch, record = journal, remainingPhaseMs = 900_000 } = {}) {
  let noEgress = false;
  let lastStream;
  if (!Number.isSafeInteger(remainingPhaseMs) || remainingPhaseMs <= 0 || remainingPhaseMs > 900_000) {
    throw new Error('invalid remaining phase timeout');
  }
  const deadline = AbortSignal.timeout(remainingPhaseMs);
  const guarded = async (input, init = {}) => {
    const url = String(input); const method = (init.method || 'GET').toUpperCase();
    const headers = new Headers(init.headers);
    const body = init.body ? JSON.parse(init.body) : undefined;
    if (method === 'POST' && !url.endsWith('/v1/route')) {
      if (!headers.has('x-conifer-max-cost-nanousd')) {
        headers.set('x-conifer-max-cost-nanousd', String(body?.model === 'auto' ? 350_000_000 : 50_000_000));
      }
      headers.set('x-conifer-cache', 'off');
    }
    const admission = record('admit', { url, method, body, no_egress: noEgress, case_name: guarded.caseName,
      headers: Object.fromEntries([...headers].filter(([key]) => key !== 'authorization')) });
    headers.set('idempotency-key', admission.request_id);
    const timeout = AbortSignal.timeout(url.endsWith('/v1/route') ? 15_000 : 90_000);
    const signals = [deadline, timeout, ...(init.signal ? [init.signal] : [])];
    try {
      const response = await dispatch(url, { ...init, method, headers, redirect: 'error', signal: AbortSignal.any(signals) });
      const stream = response.ok && response.headers.get('content-type')?.includes('text/event-stream');
      let parsed;
      if (!stream) {
        const text = await response.clone().text();
        try { parsed = JSON.parse(text); } catch { parsed = { non_json_body: text.slice(0, 1000) }; }
      }
      record('observe', { index: admission.index, status: response.status, headers: Object.fromEntries(response.headers), body: parsed, stream: !!stream });
      if (stream) lastStream = admission.index;
      return response;
    } catch (error) {
      record('fault', { index: admission.index, message: String(error.message || error) });
      throw error;
    }
  };
  guarded.withoutEgress = async (run) => {
    noEgress = true;
    try { return await run(); } finally { noEgress = false; }
  };
  guarded.streamDone = (usage) => record('stream', { index: lastStream, usage });
  return guarded;
}

File: scripts/live_qa_transport.py
"""Wrap the real urllib boundary, including SDK streaming, for bounded QA."""
from __future__ import annotations
import io
import json
import math
import urllib.error
import urllib.request
from live_qa_guard import BASE_URL, BUILD, Guard, GuardError

PINS = {'chat': 'gpt-4.1-nano', 'spare': 'gpt-4.1-mini', 'native': 'claude-haiku-4-5', 'embed': 'text-embedding-3-small'}


def assert_model(model, cap, capability, provider=None):
    if not model or model.endpoint_kind != 'conifer' or model.unavailable is True or (capability and capability not in (model.caps or [])) or (provider and model.provider != provider):
        raise AssertionError('fixed model absent, unavailable or incompatible with wire')
    for field in ('in_usd_per_mtok', 'out_usd_per_mtok'):
        price = (model.pricing or {}).get(field)
        if not isinstance(price, str) or not math.isfinite(float(price)) or float(price) < 0:
            raise AssertionError('unpriced model')
    if cap is not None and (model.output_token_limit_supported is False or (model.min_output_tokens or 0) > cap or (model.max_output_tokens or cap) < cap):
        raise AssertionError(f'{model.id} cannot honor {cap} output tokens')
    return model


def empty_outcome(error):
    if (getattr(error, 'status', None) != 422 or error.type != 'output_budget_exhausted' or
        error.code != 'output_budget_exhausted' or error.param != 'max_tokens' or error.retryable is not False or
        not error.request_id or not isinstance(error.body, dict) or not error.body.get('usage')):
        raise error
    return 'typed output_budget_exhausted with preserved usage; transport verified settled receipt'


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


class BufferedResponse(io.BytesIO):
    def __init__(self, data, status, headers):
        super().__init__(data); self.status = status; self.headers = headers


class GuardedHTTP:
    def __init__(self, guard=None, opener_factory=urllib.request.build_opener):
        self.guard = guard or Guard(); self.no_egress = False; self.last_stream = None
        self.opener_factory = opener_factory

    def record(self, action, **payload):
        return self.guard.operation(action, {'phase': 'python', **payload})

    def urlopen(self, request, timeout=90, context=None, **kwargs):
        if not isinstance(request, urllib.request.Request):
            request = urllib.request.Request(request)
        url = request.full_url; method = request.get_method()
        body = json.loads(request.data) if request.data else None
        headers = {k.lower(): v for k, v in request.header_items()}
        if method == 'POST' and not url.endswith('/v1/route'):
            headers.setdefault('x-conifer-max-cost-nanousd', '50000000')
            headers['x-conifer-cache'] = 'off'
        admission = self.record('admit', url=url, method=method, body=body, no_egress=self.no_egress,
                                case_name=getattr(self, 'case_name', None),
                                headers={k: v for k, v in headers.items() if k != 'authorization'})
        headers['idempotency-key'] = admission['request_id']
        request = urllib.request.Request(url, data=request.data, headers=headers, method=method)
        opener = self.opener_factory(urllib.request.HTTPSHandler(context=context), NoRedirect())
        error_response = None
        try:
            try:
                response = opener.open(request, timeout=min(timeout, 90))
            except urllib.error.HTTPError as error:
                response = error; error_response = error
            response_headers = dict(response.headers.items())
            stream = 200 <= response.status < 300 and 'text/event-stream' in response.headers.get('content-type', '')
            data = None if stream else response.read()
            parsed = None
            if data is not None:
                try: parsed = json.loads(data)
                except (ValueError, UnicodeError): parsed = {'non_json_body': data[:1000].decode(errors='replace')}
            self.record('observe', index=admission['index'], status=response.status,
                        headers=response_headers, body=parsed, stream=stream)
            if stream:
                self.last_stream = admission['index']
                return response
            response.close()
            if error_response is not None:
                # Preserve the real HTTPError shape and its readable body for SDK error_from.
                error_response.fp = io.BytesIO(data)
                error_response.file = error_response.fp
                error_response.read = error_response.fp.read
                raise error_response
            return BufferedResponse(data, response.status, response.headers)
        except urllib.error.HTTPError:
            raise
        except Exception as error:
            self.record('fault', index=admission['index'], message=str(error))
            raise

    def without_egress(self, run):
        self.no_egress = True
        try: return run()
        finally: self.no_egress = False

    def stream_done(self, usage):
        return self.record('stream', index=self.last_stream, usage=usage)
